### PR TITLE
Entity@ initialization in script

### PR DESCRIPTION
**Pull request by Sajdkick**
Fixed some issues with the reflection of entity@ in the editor. You can now initialize an Entity@ in the script and it will show up in the editor, and if you try to play without initializing it gives you an error.

### CODE REVIEW

**Comment by NinjaDanz3r** [comment]
SonarQube analysis reported 6 issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] AncientFungus:LGEPR: Cppcheck cannot find all the include files (use --check-config for details) [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AmissingInclude)
1. ![MINOR][MINOR] [ScriptManager.cpp#L115](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/1d6e728f75a1d090410edf0946c68d5e78e32744/src/Engine/Manager/ScriptManager.cpp#L115): Variable 'callstack' is assigned a value that is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunreadVariable)
1. ![MINOR][MINOR] [ScriptManager.cpp#L116](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/1d6e728f75a1d090410edf0946c68d5e78e32744/src/Engine/Manager/ScriptManager.cpp#L116): Variable 'variables' is assigned a value that is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunreadVariable)
1. ![MINOR][MINOR] [ScriptManager.cpp#L605](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/1d6e728f75a1d090410edf0946c68d5e78e32744/src/Engine/Manager/ScriptManager.cpp#L605): The function 'ClearBreakpoints' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)
1. ![MINOR][MINOR] [ScriptManager.cpp#L847](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/1d6e728f75a1d090410edf0946c68d5e78e32744/src/Engine/Manager/ScriptManager.cpp#L847): The function 'GetStringDeclarationID' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)
1. ![MINOR][MINOR] [ScriptManager.cpp#L853](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/1d6e728f75a1d090410edf0946c68d5e78e32744/src/Engine/Manager/ScriptManager.cpp#L853): The function 'GetScripts' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by NinjaDanz3r** [comment]
SonarQube analysis reported 6 issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] AncientFungus:LGEPR: Cppcheck cannot find all the include files (use --check-config for details) [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AmissingInclude)
1. ![MINOR][MINOR] [ScriptManager.cpp#L115](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/1f33724f1b1182de1ce8c9b9fea328469bf3d68a/src/Engine/Manager/ScriptManager.cpp#L115): Variable 'callstack' is assigned a value that is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunreadVariable)
1. ![MINOR][MINOR] [ScriptManager.cpp#L116](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/1f33724f1b1182de1ce8c9b9fea328469bf3d68a/src/Engine/Manager/ScriptManager.cpp#L116): Variable 'variables' is assigned a value that is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunreadVariable)
1. ![MINOR][MINOR] [ScriptManager.cpp#L609](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/1f33724f1b1182de1ce8c9b9fea328469bf3d68a/src/Engine/Manager/ScriptManager.cpp#L609): The function 'ClearBreakpoints' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)
1. ![MINOR][MINOR] [ScriptManager.cpp#L851](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/1f33724f1b1182de1ce8c9b9fea328469bf3d68a/src/Engine/Manager/ScriptManager.cpp#L851): The function 'GetStringDeclarationID' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)
1. ![MINOR][MINOR] [ScriptManager.cpp#L857](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/1f33724f1b1182de1ce8c9b9fea328469bf3d68a/src/Engine/Manager/ScriptManager.cpp#L857): The function 'GetScripts' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by NinjaDanz3r** [comment]
SonarQube analysis reported 6 issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] AncientFungus:LGEPR: Cppcheck cannot find all the include files (use --check-config for details) [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AmissingInclude)
1. ![MINOR][MINOR] [ScriptManager.cpp#L115](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/60dc2be44706ce01bc5b11b23e893a2e4cc97e13/src/Engine/Manager/ScriptManager.cpp#L115): Variable 'callstack' is assigned a value that is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunreadVariable)
1. ![MINOR][MINOR] [ScriptManager.cpp#L116](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/60dc2be44706ce01bc5b11b23e893a2e4cc97e13/src/Engine/Manager/ScriptManager.cpp#L116): Variable 'variables' is assigned a value that is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunreadVariable)
1. ![MINOR][MINOR] [ScriptManager.cpp#L609](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/60dc2be44706ce01bc5b11b23e893a2e4cc97e13/src/Engine/Manager/ScriptManager.cpp#L609): The function 'ClearBreakpoints' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)
1. ![MINOR][MINOR] [ScriptManager.cpp#L852](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/60dc2be44706ce01bc5b11b23e893a2e4cc97e13/src/Engine/Manager/ScriptManager.cpp#L852): The function 'GetStringDeclarationID' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)
1. ![MINOR][MINOR] [ScriptManager.cpp#L858](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/60dc2be44706ce01bc5b11b23e893a2e4cc97e13/src/Engine/Manager/ScriptManager.cpp#L858): The function 'GetScripts' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by NinjaDanz3r** [comment]
SonarQube analysis reported 6 issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] AncientFungus:LGEPR: Cppcheck cannot find all the include files (use --check-config for details) [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AmissingInclude)
1. ![MINOR][MINOR] [ScriptManager.cpp#L115](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/52292682c4405eeb1b6d12d8507a942bc8d9708d/src/Engine/Manager/ScriptManager.cpp#L115): Variable 'callstack' is assigned a value that is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunreadVariable)
1. ![MINOR][MINOR] [ScriptManager.cpp#L116](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/52292682c4405eeb1b6d12d8507a942bc8d9708d/src/Engine/Manager/ScriptManager.cpp#L116): Variable 'variables' is assigned a value that is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunreadVariable)
1. ![MINOR][MINOR] [ScriptManager.cpp#L609](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/52292682c4405eeb1b6d12d8507a942bc8d9708d/src/Engine/Manager/ScriptManager.cpp#L609): The function 'ClearBreakpoints' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)
1. ![MINOR][MINOR] [ScriptManager.cpp#L852](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/52292682c4405eeb1b6d12d8507a942bc8d9708d/src/Engine/Manager/ScriptManager.cpp#L852): The function 'GetStringDeclarationID' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)
1. ![MINOR][MINOR] [ScriptManager.cpp#L858](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/52292682c4405eeb1b6d12d8507a942bc8d9708d/src/Engine/Manager/ScriptManager.cpp#L858): The function 'GetScripts' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: src/Editor/GUI/Editors/EntityEditor.cpp [hunk 1]
```diff
@@ -473,21 +473,45 @@ void EntityEditor::ScriptEditor(Component::Script* script) {
                 else if (typeId == script->instance->GetEngine()->GetTypeIdByDecl("Entity@")) {
 
                     // Find method to call.
-                    std::string entityName = script->instance->GetPropertyName(n);
+                    std::string propertyName = script->instance->GetPropertyName(n);
```
Comment (andens) [inline src/Editor/GUI/Editors/EntityEditor.cpp]: A variable with the same value and name already exists in the beginning of the for loop. You may want to consider removing this line altogether.

File: src/Editor/GUI/Editors/EntityEditor.cpp [hunk 2]
```diff
@@ -473,21 +473,45 @@ void EntityEditor::ScriptEditor(Component::Script* script) {
...
-                        propertyText.append(entityName).append(": ").append(entityGUID);
+                        unsigned int GUID = *(unsigned int*)script->GetDataFromPropertyMap(propertyName);
 
-                        ImGui::Separator();
+                        if (GUID != 0) {
+
+                            std::string entityGUID = std::to_string(GUID);
+                            std::string entityName = Hymn().GetEntityByGUID(GUID)->name;
+                            std::string propertyText;
+                            propertyText.reserve(propertyName.length() + entityName.length() + entityGUID.length() + 2); // additional `:  ()`
```
Comment (andens) [inline src/Editor/GUI/Editors/EntityEditor.cpp]: Not a biggie, but the addition of 2 should probably be 4 to accomodate for the extra parantheses now added.

File: src/Editor/GUI/Editors/EntityEditor.cpp [hunk 3]
```diff
@@ -473,21 +473,45 @@ void EntityEditor::ScriptEditor(Component::Script* script) {
...
+                        } else {
 
-                        // Choosing other entity references
-                        ImGui::Text(propertyText.c_str());
-                        if (ImGui::Button("Change entity reference"))
-                            ImGui::OpenPopup("Add entity reference");
+                            std::string entityGUID = "???";
+                            std::string entityName = "Uninitialized";
+                            std::string propertyText;
+                            propertyText.reserve(propertyName.length() + entityName.length() + entityGUID.length() + 2); // additional `:  ()`
```
Comment (andens) [inline src/Editor/GUI/Editors/EntityEditor.cpp]: The number 2 should be 4 here as well I think.

File: src/Engine/Manager/ScriptManager.cpp [hunk 4]
```diff
@@ -703,9 +724,18 @@ void ScriptManager::Update(World& world, float deltaTime) {
...
+                        unsigned int* GUID = (unsigned int*)script->GetDataFromPropertyMap(name);
+
+                        //We make sure it is initialized.
+                        if (*GUID != 0)
+                            *reinterpret_cast<Entity**>(varPointer) = Hymn().GetEntityByGUID(*GUID);
+                        else
+                            Log() << "Property " << name << " of script " << script->scriptFile->name << " on entity " << script->entity->name << " is not initialized" << "\n";
+
+                    }
+                    else
```
Comment (andens) [inline src/Engine/Manager/ScriptManager.cpp]: This goes on the previous line.

**Comment by NinjaDanz3r** [comment]
SonarQube analysis reported 6 issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] AncientFungus:LGEPR: Cppcheck cannot find all the include files (use --check-config for details) [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AmissingInclude)
1. ![MINOR][MINOR] [ScriptManager.cpp#L115](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/a76fe1a5a14ac287895cbd18c5e73f471e2c52d3/src/Engine/Manager/ScriptManager.cpp#L115): Variable 'callstack' is assigned a value that is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunreadVariable)
1. ![MINOR][MINOR] [ScriptManager.cpp#L116](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/a76fe1a5a14ac287895cbd18c5e73f471e2c52d3/src/Engine/Manager/ScriptManager.cpp#L116): Variable 'variables' is assigned a value that is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunreadVariable)
1. ![MINOR][MINOR] [ScriptManager.cpp#L609](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/a76fe1a5a14ac287895cbd18c5e73f471e2c52d3/src/Engine/Manager/ScriptManager.cpp#L609): The function 'ClearBreakpoints' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)
1. ![MINOR][MINOR] [ScriptManager.cpp#L851](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/a76fe1a5a14ac287895cbd18c5e73f471e2c52d3/src/Engine/Manager/ScriptManager.cpp#L851): The function 'GetStringDeclarationID' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)
1. ![MINOR][MINOR] [ScriptManager.cpp#L857](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/a76fe1a5a14ac287895cbd18c5e73f471e2c52d3/src/Engine/Manager/ScriptManager.cpp#L857): The function 'GetScripts' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by NinjaDanz3r** [comment]
SonarQube analysis reported 6 issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] AncientFungus:LGEPR: Cppcheck cannot find all the include files (use --check-config for details) [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AmissingInclude)
1. ![MINOR][MINOR] [ScriptManager.cpp#L115](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/d2ea8e75853ea73a1b786490e3b5b4b78e098643/src/Engine/Manager/ScriptManager.cpp#L115): Variable 'callstack' is assigned a value that is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunreadVariable)
1. ![MINOR][MINOR] [ScriptManager.cpp#L116](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/d2ea8e75853ea73a1b786490e3b5b4b78e098643/src/Engine/Manager/ScriptManager.cpp#L116): Variable 'variables' is assigned a value that is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunreadVariable)
1. ![MINOR][MINOR] [ScriptManager.cpp#L609](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/d2ea8e75853ea73a1b786490e3b5b4b78e098643/src/Engine/Manager/ScriptManager.cpp#L609): The function 'ClearBreakpoints' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)
1. ![MINOR][MINOR] [ScriptManager.cpp#L851](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/d2ea8e75853ea73a1b786490e3b5b4b78e098643/src/Engine/Manager/ScriptManager.cpp#L851): The function 'GetStringDeclarationID' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)
1. ![MINOR][MINOR] [ScriptManager.cpp#L857](https://github.com/Chainsawkitten/LargeGameProjectEngine/blob/d2ea8e75853ea73a1b786490e3b5b4b78e098643/src/Engine/Manager/ScriptManager.cpp#L857): The function 'GetScripts' is never used. [rule](http://localhost:9000/coding_rules#rule_key=cppcheck%3AunusedFunction)

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Pull request merged**